keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,620 @@
+`v1alpha1` and resource `authenticationconfigs`.  The `AuthenticationConfig`
+`Kind` is defined as:
+
+```golang
+type AuthenticationConfig struct {
+  metav1.TypeMeta `json:",inline"`
+  // +optional
+  metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+  Spec AuthenticationConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+type AuthenticationConfigSpec struct {
+  Type AuthenticationConfigType `json:"type" protobuf:"bytes,1,opt,name=type"`
+
+  // this generic username and group prefix config is always required
+  // but the fields can be explicitly set to "-" to override it
+  Prefix *PrefixConfig `json:"prefix" protobuf:"bytes,2,opt,name=prefix"`
+
+  X509 *X509Config `json:"x509,omitempty" protobuf:"bytes,3,opt,name=x509"`
+
+  OIDC *OIDCConfig `json:"oidc,omitempty" protobuf:"bytes,4,opt,name=oidc"`
+
+  // This struct will be defined in a later revision
+  // Webhook *WebhookConfig `json:"webhook,omitempty" protobuf:"bytes,5,opt,name=webhook"`
+}
+
+type AuthenticationConfigType string
+
+const (
+  AuthenticationConfigTypeX509    AuthenticationConfigType = "x509"
+  AuthenticationConfigTypeOIDC    AuthenticationConfigType = "oidc"
+  AuthenticationConfigTypeWebhook AuthenticationConfigType = "webhook"
+)
+
+type X509Config struct {
+  // caBundle is a PEM encoded CA bundle used for client auth (x509.ExtKeyUsageClientAuth).
+  // +listType=atomic
+  // Required
+  CABundle []byte `json:"caBundle" protobuf:"bytes,1,opt,name=caBundle"`
+}
+
+type OIDCConfig struct {
+  // this type is defined in admission registration
+  Issuer admissionregistrationv1.WebhookClientConfig `json:"issuer" protobuf:"bytes,1,opt,name=issuer"`
+
+  ClientID string `json:"clientID" protobuf:"bytes,2,opt,name=clientID"`
+
+  UsernameClaim string `json:"usernameClaim" protobuf:"bytes,3,opt,name=usernameClaim"`
+
+  GroupsClaim string `json:"groupsClaim" protobuf:"bytes,4,opt,name=groupsClaim"`
+}
+
+type PrefixConfig struct {
+  UsernamePrefix string `json:"usernamePrefix" protobuf:"bytes,1,opt,name=usernamePrefix"`
+
+  GroupsPrefix string `json:"groupsPrefix" protobuf:"bytes,2,opt,name=groupsPrefix"`
+}
+```
+
+The `spec.type` field serves as a union discriminator to determine which config
+struct to read.
+
+Most of the functionality required to create the dynamic authenticator already
+exists in the API server (to read and process the CLI flags).  It simply needs
+to be wired up to an informer that triggers a rebuild of an authenticator that
+will be stored in an `atomic.Value`.
+
+A new `authenticatorfactory.CAContentProvider` will be required to support
+dynamic certificate based authentication on the API server (via
+`BuiltInAuthenticationOptions`) and the kubelet (via `DelegatingAuthenticatorConfig`).
+This `CAContentProvider` will look similar to the dynamic authenticator except
+that it will build and maintain a concatenation of all of the CA bundles from
```

Comment:
Erm. How do we provide ordering here if they're all just smooshed together?  Particularly wrt flags vs dynamic config?

In particular, a broken dynamic x509 provider should not be able to affect the cli-configured x509 provider that you need to use to _fix_ the broken authconfig.

Reply:
> Erm. How do we provide ordering here if they're all just smooshed together? Particularly wrt flags vs dynamic config?

All the dynamic stuff is always after the static stuff.  Thus the flags would always go first.

> In particular, a broken dynamic x509 provider should not be able to affect the cli-configured x509 provider that you need to use to _fix_ the broken authconfig.

You cannot have a broken `x509` provider because the config can be statically validated.  Even without that, invalid PEM bundles would just be ignored.  There is no ordering requirements for CA bundles - any bundle that verifies the cert is fine.  The identity comes from the cert so the CA bundles themselves do not matter.